infra/bisector.py

```diff
@@ -0,0 +1,175 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Uses _bisection to determine which commit a bug was introduced and fixed.
+This module takes a high and a low commit SHA, a repo name, and a bug.
+The module bisects the high and low commit SHA searching for the location
+where the bug was introduced. It also looks for where the bug was fixed.
+This is done with the following steps:
+
+
+  NOTE: NEEDS TO BE RUN FROM THE oss-fuzz HOME directory
+
+  Typical usage example:
+        python3 infra/bisector.py
+          --commit_old 1e403e9259a1abedf108ab86f711ba52c907226d
+          --commit_new f79be4f2330f4b89ea2f42e1c44ca998c59a0c0f
+          --fuzz_target rules_fuzzer
+          --project_name yara
+          --test_case infra/yara_test_case
+          --sanitizer address
+"""
+
+import argparse
+import os
+
+from build_specified_commit import build_fuzzer_from_commit
+from build_specified_commit import infer_main_repo
+from helper import reproduce_impl
+from RepoManager import RepoManager
+
+
+def main():
+  """Finds the commit SHA where an error was initally introduced."""
+  parser = argparse.ArgumentParser(
+      'bisector.py',
+      description='git bisection for finding introduction of bugs')
+
+  parser.add_argument(
+      '--project_name',
+      help='The name of the project where the bug occured',
+      required=True)
+  parser.add_argument(
+      '--commit_new',
+      help='The newest commit SHA to be bisected',
+      required=True)
+  parser.add_argument(
+      '--commit_old',
+      help='The oldest commit SHA to be bisected',
+      required=True)
+  parser.add_argument(
+      '--fuzz_target', help='the name of the fuzzer to be built', required=True)
+  parser.add_argument(
+      '--test_case', help='the test_case to be reproduced', required=True)
+  parser.add_argument('--engine', default='libfuzzer')
+  parser.add_argument(
+      '--sanitizer',
+      default='address',
+      help='the default is "address"')
+  parser.add_argument('--architecture', default='x86_64')
+  args = parser.parse_args()
+
+  if not os.getcwd().endswith('oss-fuzz'):
+    print("Error: bisector.py needs to be run from the oss-fuzz home directory")
+    return 1
+
+  error_sha = bisect(args.project_name, args.commit_old,
+                             args.commit_new, args.engine, args.sanitizer,
+                             args.architecture, args.test_case,
+                             args.fuzz_target)
+  if not error_sha:
+    print('No error was found in commit range %s:%s' %
+          (args.commit_old, args.commit_new))
+  else:
+    print('Error was introduced at commit %s' % error_sha)
+
+
+def bisect(project_name, commit_old, commit_new, engine, sanitizer,
+                   architecture, test_case, fuzz_target):
+  """Creates an bisection call that kicks off the error detection.
+
+  This function is necessary in order to get the error code of the newest commit. This
```

Comment:
Is this actually necessary? Does simply checking for nonzero break? I think comparing the error codes adds complexity that I'm unsure is worth it.
@Dor1s WDYT?

Reply:
I believe that it is necessary. There are a lot of cases where a second error exists through the whole regression range, making the bisector thing the error was introduced through the full range.